src/Raven.Server/Documents/Indexes/Static/V8/JavaScriptMapOperation.cs

```diff
@@ -0,0 +1,143 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+using Esprima.Ast;
+using Jint;
+using Jint.Native;
+using Jint.Native.Array;
+using Jint.Native.Function;
+using Jint.Runtime;
+using Jint.Runtime.Environments;
+using Raven.Server.Extensions.Jint;
+
+using V8.Net;
+using Raven.Client.Documents.Indexes;
+using Raven.Server.Documents.Patch.V8;
+using Raven.Server.Extensions.V8;
+using Raven.Server.Documents.Indexes.Static.Utils;
+using Raven.Server.Documents.Patch;
+
+namespace Raven.Server.Documents.Indexes.Static.V8
+{
+    public class JavaScriptMapOperation : JavaScriptMapOperationBase
+    {
+        private readonly JavaScriptIndexUtils _javaScriptIndexUtils;
+        private readonly V8EngineEx _engine;
+        public InternalHandle MapFunc;
+
+        public JavaScriptMapOperation(JavaScriptIndexUtils javaScriptIndexUtils, FunctionInstance mapFuncJint, InternalHandle mapFunc, string indexName, string mapString)
+            : base((Engine)javaScriptIndexUtils.EngineJintForV8)
+        {
+            _javaScriptIndexUtils = javaScriptIndexUtils;
+            _engine = (V8EngineEx)_javaScriptIndexUtils.EngineHandle;
+
+            InternalHandle mapFuncV8Aux = mapFunc;
+            MapFunc = new InternalHandle(ref mapFuncV8Aux, true);
+            MapString = mapString;
+        }
+
+        ~JavaScriptMapOperation()
+        {
+            MapFunc.Dispose();
+        }
+
+        public override IEnumerable<InternalHandle> IndexingFunction(IEnumerable<object> items)
+        {
+            foreach (var item in items)
+            {
+                _engine.ResetCallStack();
+                _engine.ResetConstraints();
+
+                if (_javaScriptIndexUtils.GetValue(item, out JsHandle jsItemHandle) == false)
+                {
+                    continue;
+                }
+
+#if DEBUG
+                _engine.MakeSnapshot("map");
```

Comment:
what is that for?

Reply:
for JS entities (InternalHandle) leakage detection (is performed for map and reduce separately in debug version)